rmw_connext_cpp/src/publisher.cpp

```diff
@@ -0,0 +1,376 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rmw_connext_cpp/publisher.hpp"
+
+#include <string>
+
+#include "rmw/allocators.h"
+#include "rmw/error_handling.h"
+#include "rmw/impl/cpp/macros.hpp"
+#include "rmw/rmw.h"
+#include "rmw/types.h"
+
+#include "rmw_connext_shared_cpp/qos.hpp"
+#include "rmw_connext_shared_cpp/types.hpp"
+
+#include "rmw_connext_cpp/connext_static_publisher_info.hpp"
+#include "rmw_connext_cpp/identifier.hpp"
+
+// include patched generated code from the build folder
+#include "connext_static_serialized_dataSupport.h"
+
+#include "process_topic_and_service_names.hpp"
+#include "type_support_common.hpp"
+
+// Uncomment this to get extra console output about discovery.
+// This affects code in this file, but there is a similar variable in:
+//   rmw_connext_shared_cpp/shared_functions.cpp
+// #define DISCOVERY_DEBUG_LOGGING 1
+
+rmw_publisher_t *
+rmw_connext_cpp::create_publisher(
+  const ConnextParticipantInfo * participant_info,
+  const rosidl_message_type_support_t * type_supports,
+  const char * topic_name,
+  const rmw_qos_profile_t * qos_profile,
+  const rmw_publisher_options_t * publisher_options)
+{
+  RMW_CONNEXT_EXTRACT_MESSAGE_TYPESUPPORT(type_supports, type_support, NULL)
+
+  if (!topic_name || strlen(topic_name) == 0) {
+    RMW_SET_ERROR_MSG("publisher topic is null or empty string");
+    return NULL;
+  }
+
+  if (!qos_profile) {
+    RMW_SET_ERROR_MSG("qos_profile is null");
+    return NULL;
+  }
+
+  if (!publisher_options) {
+    RMW_SET_ERROR_MSG("publisher_options is null");
+    return NULL;
+  }
+
+  if (!participant_info) {
+    RMW_SET_ERROR_MSG("participant_info is null");
+    return NULL;
+  }
+  auto participant = static_cast<DDS::DomainParticipant *>(participant_info->participant);
+  if (!participant) {
+    RMW_SET_ERROR_MSG("participant handle is null");
+    return NULL;
+  }
+
+  const message_type_support_callbacks_t * callbacks =
+    static_cast<const message_type_support_callbacks_t *>(type_support->data);
+  if (!callbacks) {
+    RMW_SET_ERROR_MSG("callbacks handle is null");
+    return NULL;
+  }
+  std::string type_name = _create_type_name(callbacks);
+  // Past this point, a failure results in unrolling code in the goto fail block.
+  DDS::TypeCode * type_code = nullptr;
+  DDS::DataWriterQos datawriter_qos;
+  DDS::PublisherQos publisher_qos;
+  DDS::ReturnCode_t status;
+  DDS::Publisher * dds_publisher = nullptr;
+  DDS::DataWriter * topic_writer = nullptr;
+  DDS::Topic * topic = nullptr;
+  DDS::TopicDescription * topic_description = nullptr;
+  void * info_buf = nullptr;
+  void * listener_buf = nullptr;
+  ConnextPublisherListener * publisher_listener = nullptr;
+  ConnextStaticPublisherInfo * publisher_info = nullptr;
+  rmw_publisher_t * publisher = nullptr;
+  std::string mangled_name = "";
```

Comment:
@ivanpauno nit: no need to initialize with an empty string.

Reply:
I actually copy-pasted that, but ok.

Reply:
Addressed in https://github.com/ros2/rmw_connext/pull/382/commits/4e2e037593cf27fa1e1201bb333fb2e7680a1971.